core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,337 @@
+    }
+    actionsAfterEOF();
+    return false;
+  }
+
+  /**
+   * Method to append start tag information along with attributes if any
+   */
+  private void appendStartTagInformation(String nodeNameStart, StringBuilder xmlRecord) {
+    xmlRecord.append(OPENING_START_TAG_DELIMITER).append(nodeNameStart);
+    int count = reader.getAttributeCount();
+    for (int i = 0; i < count; i++) {
+      xmlRecord.append(" ").append(reader.getAttributeLocalName(i)).append("=\"").append(reader.getAttributeValue(i))
+        .append("\"");
+    }
+    xmlRecord.append(CLOSING_START_TAG_DELIMITER);
+  }
+
+  /**
+   * Method to take actions after EOF reached.
+   */
+  private void actionsAfterEOF() throws IOException {
+    if (StringUtils.isNotEmpty(fileAction)) {
+      processFileAction();
+    }
+    updateFileTrackingInfo();
+  }
+
+  /**
+   * Method to process file with actions (Delete, Move, Archive ) specified.
+   * @throws IOException - IO Exception occurred while deleting, moving or archiving file.
+   */
+  private void processFileAction() throws IOException {
+    fileAction = fileAction.toLowerCase();
+    switch (fileAction) {
+      case "delete":
+        fs.delete(file, true);
+        break;
+      case "move":
+        Path tagetFileMovePath = new Path(targetFolder + file.getName());
+        fs.rename(file, tagetFileMovePath);
+        break;
+      case "archive":
+        FileOutputStream archivedStream = new FileOutputStream(targetFolder + file.getName() + ".zip");
+        ZipOutputStream zipArchivedStream = new ZipOutputStream(archivedStream);
+        FSDataInputStream fdDataInputStream = null;
+        try {
+          fdDataInputStream = fs.open(file);
+          zipArchivedStream.putNextEntry(new ZipEntry(file.getName()));
+          int length;
+          byte[] buffer = new byte[1024];
+          while ((length = fdDataInputStream.read(buffer)) > 0) {
+            zipArchivedStream.write(buffer, 0, length);
+          }
+        } catch (IOException ioException) {
+          throw ioException;
+        } finally {
+          if (zipArchivedStream != null) {
+            zipArchivedStream.closeEntry();
+          }
+          fdDataInputStream.close();
+          zipArchivedStream.close();
+        }
+        fs.delete(file, true);
+        break;
+      default:
+        LOG.warn("No action required on the file.");
+        break;
+    }
+  }
+
+  /**
+   * Method to update temporary file with latest XML processed information.
+   * @throws IOException - IO Exception occurred while writing data to temp file.
+   */
+  private void updateFileTrackingInfo() throws IOException {
+    //TODO - remove temp file usage after proper solution to send data back to XMLReaderBatchSource.
+    File tempFile = new File(tempFilePath);
+    if (!tempFile.exists()) {
+      tempFile.createNewFile();
+    }
+    FileWriter fw = new FileWriter(tempFile.getAbsoluteFile(), true);
+    BufferedWriter bw = new BufferedWriter(fw);
```

Comment:
use try-catch block for BufferedWriter, to ensure it gets closed.
(also no longer need to close it explicitly).